### PR TITLE
Drop support for LLVM8 (Issue #4560)

### PR DESCRIPTION
**Pull request by steven-johnson**
Drop LLVM8 support (in anticipation of LLVM10 being tagged later this week).

### CODE REVIEW

File: src/CodeGen_LLVM.cpp [hunk 1]
```diff
@@ -1320,110 +1318,6 @@ void CodeGen_LLVM::optimize_module() {
...
-    module_pass_manager.add(createTargetTransformInfoWrapperPass(tm ? tm->getTargetIRAnalysis() : TargetIRAnalysis()));
-    function_pass_manager.add(createTargetTransformInfoWrapperPass(tm ? tm->getTargetIRAnalysis() : TargetIRAnalysis()));
-
-    PassManagerBuilder b;
-    b.OptLevel = 3;
-    b.Inliner = createFunctionInliningPass(b.OptLevel, 0, false);
-    b.LoopVectorize = do_loop_opt;
-    b.DisableUnrollLoops = !do_loop_opt;
-    b.SLPVectorize = true;  // Note: SLP vectorization has no analogue in the Halide scheduling model
-#if LLVM_VERSION >= 90
```
Comment (abadams) [inline src/CodeGen_LLVM.cpp]: Was this if just dead code then?

File: test/correctness/atomics.cpp [hunk 2]
```diff
@@ -1236,14 +1236,12 @@ int main(int argc, char **argv) {
         test_all<uint64_t>(Backend::CUDA);
         test_all<int64_t>(Backend::CUDA);
         test_all<double>(Backend::CUDA);
-#if LLVM_VERSION >= 90
         test_all<uint32_t>(Backend::CUDAVectorize);
```
Comment (abadams) [inline test/correctness/atomics.cpp]: This is going to cause build failures. In master LLVM_VERSION is not defined for the tests, so these tests silently don't run. However these tests are also broken. I have fixed this in the atomic_vectorization branch

File: src/CodeGen_LLVM.cpp (hunk 1 above)
Comment (steven-johnson) [reply src/CodeGen_LLVM.cpp]: Yeah, it got nested by an outer ifdef and never cleaned up, apparently

File: test/correctness/atomics.cpp (hunk 2 above)
Comment (steven-johnson) [reply test/correctness/atomics.cpp]: ...ok then :-)

Comment (BachiLi) [reply test/correctness/atomics.cpp]: Sorry ; (

File: .github/workflows/presubmit.yml [hunk 3]
```diff
@@ -25,10 +25,10 @@ jobs:
       - uses: actions/checkout@v2
       - name: Install clang-tidy
         run: |
-          sudo apt-get install clang-tidy-9
+          sudo apt-get install llvm-9 clang-tidy-9
       - name: Build Compilation DB
         run: |
-          cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -S . -B build
+          cmake -DCMAKE_BUILD_TYPE=Debug -DCMAKE_EXPORT_COMPILE_COMMANDS=ON -DHALIDE_REQUIRE_LLVM_VERSION=90 -S . -B build
```
Comment (alexreinking) [inline .github/workflows/presubmit.yml]: Looks good to me, here. This is probably better than setting LLVM_Dir / Clang_DIR ourselves.

File: src/CodeGen_LLVM.cpp [hunk 4]
```diff
@@ -1320,110 +1318,6 @@ void CodeGen_LLVM::optimize_module() {
...
-                  i->getName().startswith("halide_cond_"))) {
-                i->addFnAttr(Attribute::SanitizeThread);
-            }
-        }
-        function_pass_manager.run(*i);
-    }
-    function_pass_manager.doFinalization();
-    module_pass_manager.run(*module);
-#endif
-
```
Comment (alexreinking) [inline src/CodeGen_LLVM.cpp]: Woo! Fun to get rid of this much compatibility code.

File: test/correctness/atomics.cpp [hunk 5]
```diff
@@ -1237,12 +1237,13 @@ int main(int argc, char **argv) {
         test_all<int64_t>(Backend::CUDA);
         test_all<double>(Backend::CUDA);
 #if LLVM_VERSION >= 90
-        test_all<uint32_t>(Backend::CUDAVectorize);
-        test_all<int32_t>(Backend::CUDAVectorize);
-        test_all<float>(Backend::CUDAVectorize);
-        test_all<uint64_t>(Backend::CUDAVectorize);
-        test_all<int64_t>(Backend::CUDAVectorize);
-        test_all<double>(Backend::CUDAVectorize);
+        // TODO: Broken currently; commented out until https://github.com/halide/Halide/pull/4628 lands
```
Comment (alexreinking) [inline test/correctness/atomics.cpp]: nit: get rid of the `#if` here.

File: src/LLVM_Headers.h [hunk 6]
```diff
@@ -1,7 +1,7 @@
 #ifndef HALIDE_LLVM_HEADERS_H
 #define HALIDE_LLVM_HEADERS_H
 
-#if LLVM_VERSION >= 80
+#if LLVM_VERSION >= 90
 // We're good to go
 #else
 #error "Compiling Halide requires LLVM 8.0 or newer"
```
Comment (dsharletg) [inline src/LLVM_Headers.h]: 8.0 -> 9.0

Comment (steven-johnson) [reply src/LLVM_Headers.h]: fixed

**Comment by steven-johnson** [comment]
FYI: this is currently waiting on LLVM10 to be finailized, and prebuilt(s) put out for Travis to use.

**Pull request merged**